src/ui_iteminfo.cpp

```diff
@@ -0,0 +1,153 @@
+                        if( i.sName == k.sName && i.sType == k.sType ) {
+                            double iVal = i.dValue;
+                            double kVal = k.dValue;
+                            if( i.sFmt != k.sFmt ) {
+                                // Different units, compare unit adjusted vals
+                                iVal = i.dUnitAdjustedVal;
+                                kVal = k.dUnitAdjustedVal;
+                            }
+                            if( iVal > kVal - .01 &&
+                                iVal < kVal + .01 ) {
+                                this_color = c_light_gray;
+                            } else if( iVal > kVal ) {
+                                if( i.bLowerIsBetter ) {
+                                    this_color = c_light_red;
+                                } else {
+                                    this_color = c_light_green;
+                                }
+                            } else if( iVal < kVal ) {
+                                if( i.bLowerIsBetter ) {
+                                    this_color = c_light_green;
+                                } else {
+                                    this_color = c_light_red;
+                                }
+                            }
+                            break;
+                        }
+                    }
+                }
+                formatted_string += colorize( i.sValue, this_color );
+            }
+            if( pos != std::string::npos ) {
+                formatted_string += sFmt.substr( pos + 5 );
+            }
+            window.draw_colored_text( formatted_string, base_color, wrap_width );
+            same_line = !i.bNewLine;
+        }
+    }
+}
+
+iteminfo_window::iteminfo_window( item_info_data &info, point pos, int width, int height,
+                                  ImGuiWindowFlags flags ) :
+    cataimgui::window( "iteminfo", flags ),
+    data( info ), pos( pos ), width( width ), height( height )
+{
+    ctxt = input_context( "default" );
+    if( info.handle_scrolling ) {
+        ctxt.register_action( "PAGE_UP" );
+        ctxt.register_action( "PAGE_DOWN" );
+    }
+    ctxt.register_action( "CONFIRM" );
+    ctxt.register_action( "QUIT" );
+    if( info.any_input ) {
+        ctxt.register_action( "ANY_INPUT" );
+    }
+
+    ctxt.set_timeout( 10 );
+}
+
+cataimgui::bounds iteminfo_window::get_bounds()
+{
+    return { static_cast<float>( str_width_to_pixels( pos.x ) ),
+             static_cast<float>( str_height_to_pixels( pos.y ) ),
+             static_cast<float>( str_width_to_pixels( width ) ),
+             static_cast<float>( str_height_to_pixels( height ) ) };
+}
+
+void iteminfo_window::draw_controls()
+{
+    draw_item_info_imgui( *this, data, width );
+
+    if( scroll != 0 ) {
+        float scroll_pos = ImGui::GetScrollY() + scroll *
+                           ImGui::GetCurrentContext()->CurrentWindow->InnerRect.GetHeight();
```

Comment:
This formula is quite a bit different than I expected. In fact, I completely misunderstood it at first. I think it would be more conventional for `scroll` to be an unsigned integer number representing the page that the user is viewing. This would eliminate the dance of setting it to 1, advancing the `ScrollY` by a page, and then setting it back to zero again.

Reply:
I'm not really happy with how scrolling works in general, so I need to revisit that. It was quite a pain figuring out how to do it programmatically in the first place.
But one reason to do it like this and not with a number representing a page is to make the interaction between mouse and keyboard scrolling more natural.

Reply:
Ah, I see. Maybe it would be more natural if the scroll amount were in pixels, and if it were passed in to `draw_item_info_imgui` as an argument. That way the logic for keeping the value around for one frame is not split between `draw_item_info_imgui` and its caller. Or perhaps if the caller were expected to just call `ImGui::SetScrollPosY` directly.